packages/@uppy/provider-views/src/ProviderView/ProviderView.jsx

```diff
@@ -105,30 +105,54 @@ export default class ProviderView extends View {
    * @returns {Promise}   Folders/files in folder
    */
   async getFolder (id, name) {
+    const controller = new AbortController()
+    const cancelRequest = () => {
+      controller.abort()
+      this.clearSelection()
+    }
+    const getNewBreadcrumpDirectories = () => {
+      const state = this.plugin.getPluginState()
+      const index = state.directories.findIndex((dir) => id === dir.id)
+
+      if (index !== -1) {
+        return state.directories.slice(0, index + 1)
+      }
+      return state.directories.concat([{ id, title: name }])
+    }
+
+    this.plugin.uppy.on('dashboard:close-panel', cancelRequest)
+    this.plugin.uppy.on('cancel-all', cancelRequest)
     this.setLoading(true)
+
     try {
-      const res = await this.provider.list(id)
       const folders = []
       const files = []
-      let updatedDirectories
+      let path = id
 
-      const state = this.plugin.getPluginState()
-      const index = state.directories.findIndex((dir) => id === dir.id)
+      while (path) {
+        const res = await this.provider.list(path, { signal: controller.signal })
 
-      if (index !== -1) {
-        updatedDirectories = state.directories.slice(0, index + 1)
-      } else {
-        updatedDirectories = state.directories.concat([{ id, title: name }])
+        for (const f of res.items) {
+          if (f.isFolder) folders.push(f)
+          else files.push(f)
+        }
+
+        path = res.nextPagePath
+        this.nextPagePath = res.nextPagePath
+        this.username ??= res.username
```

Comment:
```suggestion
        if (res.username) this.username = res.username
```

Reply:
I don't think this is better. The username will never change during folder browsing, and it is set to `null` when the user logs out.

Reply:
How do you know it will never change? what if someone changes their username, will it not change? if not, maybe we should add a comment

Reply:
I suppose we can do it to be very sure 👍